### ISSUE TITLE
Bug: [no-deprecated] not reporting usages of deprecated declared constants as object value

### ISSUE THREAD

**Issue by json-derulo** [typescript-eslint/typescript-eslint#10487]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play#ts=5.4.3&fileType=.tsx&code=PQKhAIAEBMFMAcBOsDGBDALra4TAFBwoA2ay4KA9gHYDOG4W9AXOPYgJbUDmA3Pvip0GAWwCeAYRr1wAXkax6vcMGDgAFrHIZN4LYkqJwHWuGTxDWHFUTIUGYmMHTRYgPIAjAFZzwAb3xwcCRKeFYmDAAaFTVNcmpKPVtDY1NzS2xA8DgEVgCgoJCwhXpo1Q0tWHA0YlpEhKSDIxMzBAzoLIBffG6gA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6Jge1oBNFjpEZAIb5E3dFETRoHaJHBgAviEVA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
/** @deprecated */
declare const test: string;

const myConst = test; // here the error is reported correctly
const myObj = {
  prop: test, // here no error is reported
  deep: {
    prop: test, // here also no error is reported
  }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/no-deprecated": ["error"],
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    // ...
  }
}
```

### Expected Result

Errors are reported when deprecated declared constants are used as values in objects

### Actual Result

No errors are reported when deprecated declared constants are used as values in objects

### Additional Info

After removing the `declare` keyword, the errors are reported as expected.

**Comment by kirkwaiblinger** [comment]
@json-derulo

Yep, looks like a bug 👍

 btw you have a great GitHub screen name 😆
